### PR TITLE
⛰️ Reduce peak vram consumption with efficient selective log_softmax

### PR DESCRIPTION
**Pull request by tyler-romero**
# What does this PR do?

Many TRL Trainers use the same log_softmax -> gather operation to compute a selected set of logprobs. This approach is inefficient b/c it allocates a `bs*seqlen*vocab_size` tensor to hold the logprobs. For modest bs/seqlen/vocab_size this tensor can require >2GB vram. There are a variety of more memory efficient (and faster) approaches.

This PR creates a utility function to hold a more efficient implementation of this operation and uses that utility function broadly across TRL.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by tyler-romero** [comment]
See benchmarks here: https://github.com/huggingface/trl/pull/2773#issuecomment-2638088452

Notably, the most efficient approach in these benchmarks is not stable with bfloat16, and so we fall back to the approach that loops over log_softmax for bfloat16 and float16.

**Comment by tyler-romero** [comment]
@qgallouedec

File: examples/scripts/sft_video_llm.py [hunk 1]
```diff
@@ -50,12 +50,12 @@
 
 import requests
 import torch
-import wandb
```
Comment (tyler-romero) [inline examples/scripts/sft_video_llm.py:53]: changed by running precommit

File: tests/test_core.py [hunk 2]
```diff
@@ -44,3 +44,26 @@ def whiten(values: torch.Tensor) -> torch.Tensor:
         whiten_masked = masked_whiten(self.test_input, self.test_mask)[1:3]
         diffs = (whiten_unmasked - whiten_masked).sum()
         self.assertLess(abs(diffs.item()), 0.00001)
+
+    def test_extract_per_token_logprobs(self):
+        """Test extract_per_token_logprobs with different dtypes"""
```
Comment (tyler-romero) [inline tests/test_core.py:49]:
```suggestion
    def test_selective_log_softmax(self):
```

File: tests/test_core.py [hunk 3]
```diff
@@ -44,3 +44,25 @@ def whiten(values: torch.Tensor) -> torch.Tensor:
         whiten_masked = masked_whiten(self.test_input, self.test_mask)[1:3]
         diffs = (whiten_unmasked - whiten_masked).sum()
         self.assertLess(abs(diffs.item()), 0.00001)
+
+    def test_selective_log_softmax(self):
+        dtypes = [torch.float64, torch.float32, torch.float16, torch.bfloat16]
```
Comment (qgallouedec) [inline tests/test_core.py:49]: you can use `parametrize` here as well (see example in other tests)

**Comment by qgallouedec** [comment]
That's a super cool improvement! Thanks!
Just some minor rewarks to adresse and we're good to merge

File: trl/core.py [hunk 4]
```diff
@@ -157,3 +158,27 @@ def randn_tensor(
         latents = torch.randn(shape, generator=generator, device=rand_device, dtype=dtype, layout=layout).to(device)
 
     return latents
+
+
+def selective_log_softmax(logits, input_ids):
```
Comment (qgallouedec) [inline trl/core.py:163]: can you move this into trainer utils? as it's only use in trainers

File: trl/core.py [hunk 5]
```diff
@@ -157,3 +158,27 @@ def randn_tensor(
...
+    per_token_logps = torch.gather(logits.log_softmax(-1), dim=-1, index=labels.unsqueeze(-1)).squeeze(-1)
+    ```
+    """
+    if logits.dtype in [torch.float32, torch.float64]:
+        selected_logits = torch.gather(logits, dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)
+        # loop to reduce peak mem consumption
+        logsumexp_values = torch.stack([torch.logsumexp(lg, dim=-1) for lg in logits])
+        per_token_logps = selected_logits - logsumexp_values  # log_softmax(x_i) = x_i - logsumexp(x)
+    else:
+        # logsumexp approach is unstable with bfloat16, fall back to slightly less efficent approach
```
Comment (qgallouedec) [inline trl/core.py:177]: nice finding!

**Comment by tyler-romero** [comment]
Ready for re-review!

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
Thanks again!
